### ISSUE TITLE
The number of threads and memory available may still be wrong in distributed environments

### ISSUE THREAD

**Issue by tbeason** [duckdb/duckdb#7651]
### What happens?

PR #6550 attempted to fix a related issue (#6519), but it appears that DuckDB may still incorrectly query the core count and available memory in some instances.

### To Reproduce

I start a job on my SLURM cluster requesting 16 physical cores and 64GB RAM.

`SELECT * FROM duckdb_settings();` shows that DuckDB tries to use 128 cores and 216GB RAM

`nproc` and `python -c "import joblib; print(joblib.cpu_count())"` both return 32.

### OS:

CentOS 7

### DuckDB Version:

0.8.0

### DuckDB Client:

CLI

### Full Name:

Tyler Beason

### Affiliation:

Virginia Tech

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

**Comment by paultiq** [comment]
Looking at my K8S pod, I do see a similar issue with memory but not cores: 

DuckDB correctly reports 2 threads, in my pod with only 2 available cores on a 16 Core Node.

But, DuckDB reports 106.8GB (size of the Node, not Pod): 
`select * from duckdb_settings() where name in ('memory_limit', 'max_memory')`

![image](https://github.com/duckdb/duckdb/assets/104510378/5975849d-bfd7-4a29-8066-2eb5b6e16d06)

I have 16384MB available, as confirmed by the following:

```
import psutil

def get_memory_limit():
    with open('/sys/fs/cgroup/memory/memory.limit_in_bytes') as file:
        limit_in_bytes = int(file.read().strip())
    return limit_in_bytes
memory_limit = get_memory_limit()
memory_limit_mb = memory_limit / (1024 * 1024)  # Convert to megabytes

print(f"Memory Limit: {memory_limit_mb} MB")
```

**Comment by Tishj** [comment]
@paultiq I think that's expected as we haven't done anything to make the memory limit aware of any system imposed limitations.

The only code that's ran for this (on non-windows OS) is the following:
```c++
idx_t FileSystem::GetAvailableMemory() {
	errno = 0;
	idx_t max_memory = MinValue<idx_t>((idx_t)sysconf(_SC_PHYS_PAGES) * (idx_t)sysconf(_SC_PAGESIZE), UINTPTR_MAX);
	if (errno != 0) {
		return DConstants::INVALID_INDEX;
	}
	return max_memory;
}
```

**Comment by dirkpetersen** [comment]
I think this feature would be very useful for Slurm users 

In recent years there has been a big consolidation in HPC workload managers and Slurm is now on most new installations of research computing systems. I have been using this logic in a shell script to set environment vars for duckdb:

```
# ensure best performance on Slurm node
: "${SLURM_CPUS_ON_NODE:=8}"
if [[ -n ${SLURM_MEM_PER_NODE} ]]; then
  SLURM_MEM_PER_CPU=$(( ${SLURM_MEM_PER_NODE} / ${SLURM_CPUS_ON_NODE} ))
fi
: "${SLURM_MEM_PER_CPU:=8192}"
export DTHREADS=$(( ${SLURM_CPUS_ON_NODE} * 2 ))
export DMEM=$(( ${SLURM_MEM_PER_CPU} * ${SLURM_CPUS_ON_NODE} * 75 / 100 ))
```

and then use a PRAGMA in my *.sql file:

```
PRAGMA memory_limit='${DMEM}MB';PRAGMA threads=${DTHREADS};
```

before i expand it with envsubst:

```
duckdb < <(envsubst < hotspots.sql) --csv
```

I don't know if that is the best performance (setting threads to 2*cores)  but since this is very easy to implement I am hoping that this logic can just be incorporated in duckdb. There are many departmental Slurm clusters that are not setup with cgroups to give users only the cycles they requested so jobs can overstep their boundaries.

If cgroups are configured the memory limits are hard limits and applications that are not made aware of it can easily crash
Memory can either be set per node or per CPU and you need to check for both because Slurm will only set the environment variables the user requests and will not automatically translate NODE<->CPU which is a bit dumb

**Comment by dirkpetersen** [comment]
I still think this would be very useful and is also really easy to implement

**Comment by Tishj** [comment]
We'll be happy to review a PR @dirkpetersen

**Comment by dirkpetersen** [comment]
@Tishj , I just submitted PR https://github.com/duckdb/duckdb/pull/12978

**Comment by mst** [comment]
https://github.com/duckdb/duckdb/pull/12978 doesn't fix this
```
docker run -it --entrypoint='sh' -m=250m datacatering/duckdb:v1.1.3 -c "cat /sys/fs/cgroup/memory.max ; /duckdb -c \"select version(), current_setting('max_memory')\""
262144000
┌─────────────┬───────────────────────────────┐
│ "version"() │ current_setting('max_memory') │
│   varchar   │            varchar            │
├─────────────┼───────────────────────────────┤
│ v1.1.3      │ 7.7 GiB                       │
└─────────────┴───────────────────────────────┘
```

**Comment by paultiq** [comment]
Same as above: Doesn't work on my docker instance (using cgroup v1, above example looks like cgroup v2).

```
❯ docker run -it --entrypoint=/bin/sh -m=250m datacatering/duckdb:v1.1.3 -c "cat /sys/fs/cgroup/memory.max; cat /sys/fs/
cgroup/memory/memory.limit_in_bytes ; /duckdb -c \"select version(), current_setting('max_memory')\""
cat: /sys/fs/cgroup/memory.max: No such file or directory
262144000
┌─────────────┬───────────────────────────────┐
│ "version"() │ current_setting('max_memory') │
│   varchar   │            varchar            │
├─────────────┼───────────────────────────────┤
│ v1.1.3      │ 50.1 GiB                      │
└─────────────┴───────────────────────────────┘
```

Neither on my K8s pod: 
```
import subprocess
import duckdb

memory_max = subprocess.check_output("cat /sys/fs/cgroup/memory.max", shell=True).decode().strip()
print(f"{memory_max=} {int(memory_max)/1024/1024/1024}GiB")
duckdb.execute("select version(), current_setting('max_memory')").df()
```
memory_max='34359738368' 32.0GiB 
"version"()  |	current_setting('max_memory')
v1.1.3	   |    99.5 GiB
```

**Comment by thewisenerd** [comment]
i think i've figured out what's going wrong here

i've been testing with kubernetes environments; one with cgroupsv2 and another with cgroups; and a bunch of printf debugging

cgroupsv2 seems to work; but by reading `/sys/fs/cgroup///memory.max`, normalized to `/sys/fs/cgroup/memory.max` i guess

```
# /proc/1/cgroups
0::/
```

```
CGroups::GetMemoryLimit+
CGroups: :GetCGroupV2MemoryLimit+
CGroups: :ReadCGroupPath+
CGroups: :ReadCGroupPath-: found ::
CGroups: :GetCGroupV2MemoryLimit#: cgroup_path=/
CGroups: :GetCGroupV2MemoryLimit#: memory_max_path=/sys/fs/cgroup///memory.max
CGroups: :GetCGroupV2MemoryLimit-: reading value
CGroups::GetMemoryLimit-: cgroup_v2.limit valid
```

---

pods with cgroupsv1 looked something like this

```
CGroups::ReadCGroupPath-: did not find ::
CGroups:: GetCGroupV2MemoryLimit-: group_path is empty
CGroups: :GetCGroupV1MemoryLimit+
CGroups: :ReadMemoryCGroupPath+
CGroups: ReadMemoryCGroupPath-: Line=13:pids:/system.slice/containerd.service/kubepods-burstable-pod15d67391_a4ef_469b_a6c0_da9e2424b080.slice:cri-containerd:c84529588b5be69128b0
49a663e9c1c873eb715e9a4a782228861b7c7b79ab9a
CGroups: :ReadMemoryCGroupPath-: line=12:blkio:/system.slice/containerd.service/kubepods-burstable-pod15d67391_a4ef_469b_a6c0_da9e2424b080.slice:cri-containerd:c84529588b5be69128b
049a663e9c1c873eb715e9a4a782228861b7c7b79ab9a
CGroups:: ReadMemoryCGroupPath-: line=11:rdma:/kubepods-burstable-pod15d67391_a4ef_469b_a6c0_da9e2424b080.slice:cri-containerd:c84529588b5be69128b049a663e9c1c873eb715e9a4a78222886
1b7c7b79ab9a
CGroups:: ReadMemoryCGroupPath-: line=10:memory:/system.slice/containerd.service/kubepods-burstable-pod15d67391_a4ef_469b_a6c0_da9e2424b080.slice:cri-containerd:c84529588b5be69128
b049a663e9c1c873eb715e9a4a782228861b7c7b79ab9a
CGroups: ReadMemoryCGroupPath-: line=9:hugetlb:/kubepods-burstable-pod15d67391_a4ef_469b_a6c0_da9e2424b080.slice:cri-containerd:c84529588b5be69128b049a663e9c1c873eb715e9a4a782228
861b7c7b79ab9a
CGroups: :ReadMemoryCGroupPath-: did not find line with memory:
Cgroups::GetCGroupV1MemoryLimit-: memory_egroup_path is empty
```

cgroupsv1 does *NOT* work, and this is likely because of *two* things.

1. the line format is `hierarchy_id:controllers:path`

the current current check assumes the `memory:` will be found at pos=0, while it could be found at a non-zero pos.

https://github.com/duckdb/duckdb/blob/981c7a8573b254f8efd6b71f764c5c3bd924a53d/src/common/cgroups.cpp#L114-L116

2. the `memory_limit` path probably doesn't respect a different "fs mount" (idk what the correct cgroup term is, for root filesystem isolation)

the current logic is, find the path from the above; and plug it into `/sys/fs/cgroup/memory/%s/memory.limit_in_bytes`

https://github.com/duckdb/duckdb/blob/981c7a8573b254f8efd6b71f764c5c3bd924a53d/src/common/cgroups.cpp#L65-L66

within a kubernetes pod, which uses cgroupsv1, if i had `memory_cgroup_path` as, say `/system.slice/containerd.service/kubepods-burstable-pod9b364933_ff84_498d_947c_a52528dbc45a.slice`; 

within the *pod*, i was not able to find even the folder `/sys/fs/cgroup/memory/systemd.slice`
within the *node*, i was however able to find this (but not exactly), but at the location was at `/sys/fs/cgroup/memory/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-pod9b364933_ff84_498d_947c_a52528dbc45a.slice`; i'm assuming this is because of some containerd nuance i do not understand.

---

so basically within cgroups v1, a "fallback" path to `/sys/fs/cgroup/memory/memory.limit_in_bytes` should work; similar to when [java introduced this](https://github.com/openjdk/jdk8u-dev/commit/075d2a3987174aa719037985819ec5ccc525f7cb), however; upon some reading I also know that this value can be set to, or return "(unlimited)", and idk how to handle that properly.

---

as a workaround, one may go about setting the `SLURM_MEM_PER_NODE` i guess? but that's not a proper solve for cgroupsv1?

**Comment by Tishj** [comment]
@thewisenerd thanks for looking into this, the pointers were helpful in narrowing down the problem 👍 
I don't think `(unlimited)` will be a problem, it will just fail to resolve to a number and fall back to the system memory limit:
```c++
	// System memory detection
	auto memory = FileSystem::GetAvailableMemory();
	if (!memory.IsValid()) {
		return DBConfigOptions().maximum_memory;
	}
```

> within the pod, i was not able to find even the folder /sys/fs/cgroup/memory/systemd.slice
within the node, i was however able to find this (but not exactly), but at the location was at /sys/fs/cgroup/memory/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-pod9b364933_ff84_498d_947c_a52528dbc45a.slice; i'm assuming this is because of some containerd nuance i do not understand.

I could not replicate this, my cgroup file contained this: `2:memory:/docker/e7e72642851cc212f004efb5001eed4d3bce60da0e4e3707fb47d5a86cf79670` which resolved just fine to the right `.limit_in_bytes` file

**Comment by thewisenerd** [comment]
i did validate this by attempting to read from within the `pid=1` process on a kubernetes pod; the cgroup path does not exist 🤔 

simplest reproduction is probably with the following? (assuming forked processes within sh -c have the same cgroup)
```
# pod.yaml
apiVersion: v1
kind: Pod
metadata:
  name: alpine-cgroup-test
spec:
  containers:
  - name: alpine
    image: alpine:latest
    command: ["sh", "-c", "cat /proc/self/cgroup && ls -al /sys/fs/cgroup/memory"]
    resources:
      requests:
        memory: "64Mi"
      limits:
        memory: "128Mi"
```

with kubernetes 1.16 and docker

```
12:freezer:/kubepods/burstable/pod7de2905f-d523-45a3-8dd9-8365a286a434/610228a731e860b5e593cfdf1ae6899eb73d7c5f947e6d69716d2c3351573c90
11:net_cls,net_prio:/kubepods/burstable/pod7de2905f-d523-45a3-8dd9-8365a286a434/610228a731e860b5e593cfdf1ae6899eb73d7c5f947e6d69716d2c3351573c90
10:pids:/kubepods/burstable/pod7de2905f-d523-45a3-8dd9-8365a286a434/610228a731e860b5e593cfdf1ae6899eb73d7c5f947e6d69716d2c3351573c90
9:memory:/kubepods/burstable/pod7de2905f-d523-45a3-8dd9-8365a286a434/610228a731e860b5e593cfdf1ae6899eb73d7c5f947e6d69716d2c3351573c90
8:cpu,cpuacct:/kubepods/burstable/pod7de2905f-d523-45a3-8dd9-8365a286a434/610228a731e860b5e593cfdf1ae6899eb73d7c5f947e6d69716d2c3351573c90
7:hugetlb:/kubepods/burstable/pod7de2905f-d523-45a3-8dd9-8365a286a434/610228a731e860b5e593cfdf1ae6899eb73d7c5f947e6d69716d2c3351573c90
6:cpuset:/kubepods/burstable/pod7de2905f-d523-45a3-8dd9-8365a286a434/610228a731e860b5e593cfdf1ae6899eb73d7c5f947e6d69716d2c3351573c90
5:rdma:/
4:perf_event:/kubepods/burstable/pod7de2905f-d523-45a3-8dd9-8365a286a434/610228a731e860b5e593cfdf1ae6899eb73d7c5f947e6d69716d2c3351573c90
3:devices:/kubepods/burstable/pod7de2905f-d523-45a3-8dd9-8365a286a434/610228a731e860b5e593cfdf1ae6899eb73d7c5f947e6d69716d2c3351573c90
2:blkio:/kubepods/burstable/pod7de2905f-d523-45a3-8dd9-8365a286a434/610228a731e860b5e593cfdf1ae6899eb73d7c5f947e6d69716d2c3351573c90
1:name=systemd:/kubepods/burstable/pod7de2905f-d523-45a3-8dd9-8365a286a434/610228a731e860b5e593cfdf1ae6899eb73d7c5f947e6d69716d2c3351573c90
0::/system.slice/docker.service
total 0
drwxr-xr-x    2 root     root             0 Mar 12 02:28 .
dr-xr-xr-x   14 root     root           360 Mar 12 02:28 ..
-rw-r--r--    1 root     root             0 Mar 12 02:28 cgroup.clone_children
--w--w--w-    1 root     root             0 Mar 12 02:28 cgroup.event_control
-rw-r--r--    1 root     root             0 Mar 12 02:28 cgroup.procs
-rw-r--r--    1 root     root             0 Mar 12 02:28 memory.failcnt
--w-------    1 root     root             0 Mar 12 02:28 memory.force_empty
-rw-r--r--    1 root     root             0 Mar 12 02:28 memory.kmem.failcnt
-rw-r--r--    1 root     root             0 Mar 12 02:28 memory.kmem.limit_in_bytes
-rw-r--r--    1 root     root             0 Mar 12 02:28 memory.kmem.max_usage_in_bytes
-r--r--r--    1 root     root             0 Mar 12 02:28 memory.kmem.slabinfo
-rw-r--r--    1 root     root             0 Mar 12 02:28 memory.kmem.tcp.failcnt
-rw-r--r--    1 root     root             0 Mar 12 02:28 memory.kmem.tcp.limit_in_bytes
-rw-r--r--    1 root     root             0 Mar 12 02:28 memory.kmem.tcp.max_usage_in_bytes
-r--r--r--    1 root     root             0 Mar 12 02:28 memory.kmem.tcp.usage_in_bytes
-r--r--r--    1 root     root             0 Mar 12 02:28 memory.kmem.usage_in_bytes
-rw-r--r--    1 root     root             0 Mar 12 02:28 memory.limit_in_bytes
-rw-r--r--    1 root     root             0 Mar 12 02:28 memory.max_usage_in_bytes
-rw-r--r--    1 root     root             0 Mar 12 02:28 memory.move_charge_at_immigrate
-r--r--r--    1 root     root             0 Mar 12 02:28 memory.numa_stat
-rw-r--r--    1 root     root             0 Mar 12 02:28 memory.oom_control
----------    1 root     root             0 Mar 12 02:28 memory.pressure_level
-rw-r--r--    1 root     root             0 Mar 12 02:28 memory.soft_limit_in_bytes
-r--r--r--    1 root     root             0 Mar 12 02:28 memory.stat
-rw-r--r--    1 root     root             0 Mar 12 02:28 memory.swappiness
-r--r--r--    1 root     root             0 Mar 12 02:28 memory.usage_in_bytes
-rw-r--r--    1 root     root             0 Mar 12 02:28 memory.use_hierarchy
-rw-r--r--    1 root     root             0 Mar 12 02:28 notify_on_release
-rw-r--r--    1 root     root             0 Mar 12 02:28 tasks
```

with kubernetes 1.22 and containerd

```
13:perf_event:/kubepods-burstable-pod6a7577d9_4f2f_4403_9ec4_9401cea947e2.slice:cri-containerd:e7a42191502204647d7c041430ddd53a43679e013359f85933a5a456e2a644f5
12:memory:/system.slice/containerd.service/kubepods-burstable-pod6a7577d9_4f2f_4403_9ec4_9401cea947e2.slice:cri-containerd:e7a42191502204647d7c041430ddd53a43679e013359f85933a5a456e2a644f5
11:cpu,cpuacct:/system.slice/containerd.service/kubepods-burstable-pod6a7577d9_4f2f_4403_9ec4_9401cea947e2.slice:cri-containerd:e7a42191502204647d7c041430ddd53a43679e013359f85933a5a456e2a644f5
10:pids:/system.slice/containerd.service/kubepods-burstable-pod6a7577d9_4f2f_4403_9ec4_9401cea947e2.slice:cri-containerd:e7a42191502204647d7c041430ddd53a43679e013359f85933a5a456e2a644f5
9:rdma:/kubepods-burstable-pod6a7577d9_4f2f_4403_9ec4_9401cea947e2.slice:cri-containerd:e7a42191502204647d7c041430ddd53a43679e013359f85933a5a456e2a644f5
8:hugetlb:/kubepods-burstable-pod6a7577d9_4f2f_4403_9ec4_9401cea947e2.slice:cri-containerd:e7a42191502204647d7c041430ddd53a43679e013359f85933a5a456e2a644f5
7:misc:/kubepods-burstable-pod6a7577d9_4f2f_4403_9ec4_9401cea947e2.slice:cri-containerd:e7a42191502204647d7c041430ddd53a43679e013359f85933a5a456e2a644f5
6:net_cls,net_prio:/kubepods-burstable-pod6a7577d9_4f2f_4403_9ec4_9401cea947e2.slice:cri-containerd:e7a42191502204647d7c041430ddd53a43679e013359f85933a5a456e2a644f5
5:freezer:/kubepods-burstable-pod6a7577d9_4f2f_4403_9ec4_9401cea947e2.slice:cri-containerd:e7a42191502204647d7c041430ddd53a43679e013359f85933a5a456e2a644f5
4:cpuset:/kubepods-burstable-pod6a7577d9_4f2f_4403_9ec4_9401cea947e2.slice:cri-containerd:e7a42191502204647d7c041430ddd53a43679e013359f85933a5a456e2a644f5
3:devices:/system.slice/containerd.service/kubepods-burstable-pod6a7577d9_4f2f_4403_9ec4_9401cea947e2.slice:cri-containerd:e7a42191502204647d7c041430ddd53a43679e013359f85933a5a456e2a644f5
2:blkio:/system.slice/containerd.service/kubepods-burstable-pod6a7577d9_4f2f_4403_9ec4_9401cea947e2.slice:cri-containerd:e7a42191502204647d7c041430ddd53a43679e013359f85933a5a456e2a644f5
1:name=systemd:/system.slice/containerd.service/kubepods-burstable-pod6a7577d9_4f2f_4403_9ec4_9401cea947e2.slice:cri-containerd:e7a42191502204647d7c041430ddd53a43679e013359f85933a5a456e2a644f5
0::/kubepods-burstable-pod6a7577d9_4f2f_4403_9ec4_9401cea947e2.slice:cri-containerd:e7a42191502204647d7c041430ddd53a43679e013359f85933a5a456e2a644f5
total 0
drwxr-xr-x    2 root     root             0 Mar 12 02:28 .
dr-xr-xr-x   15 root     root           380 Mar 12 02:28 ..
-rw-r--r--    1 root     root             0 Mar 12 02:28 cgroup.clone_children
--w--w--w-    1 root     root             0 Mar 12 02:28 cgroup.event_control
-rw-r--r--    1 root     root             0 Mar 12 02:28 cgroup.procs
-rw-r--r--    1 root     root             0 Mar 12 02:28 memory.failcnt
--w-------    1 root     root             0 Mar 12 02:28 memory.force_empty
-rw-r--r--    1 root     root             0 Mar 12 02:28 memory.kmem.failcnt
-rw-r--r--    1 root     root             0 Mar 12 02:28 memory.kmem.limit_in_bytes
-rw-r--r--    1 root     root             0 Mar 12 02:28 memory.kmem.max_usage_in_bytes
-r--r--r--    1 root     root             0 Mar 12 02:28 memory.kmem.slabinfo
-rw-r--r--    1 root     root             0 Mar 12 02:28 memory.kmem.tcp.failcnt
-rw-r--r--    1 root     root             0 Mar 12 02:28 memory.kmem.tcp.limit_in_bytes
-rw-r--r--    1 root     root             0 Mar 12 02:28 memory.kmem.tcp.max_usage_in_bytes
-r--r--r--    1 root     root             0 Mar 12 02:28 memory.kmem.tcp.usage_in_bytes
-r--r--r--    1 root     root             0 Mar 12 02:28 memory.kmem.usage_in_bytes
-rw-r--r--    1 root     root             0 Mar 12 02:28 memory.limit_in_bytes
-rw-r--r--    1 root     root             0 Mar 12 02:28 memory.max_usage_in_bytes
-rw-r--r--    1 root     root             0 Mar 12 02:28 memory.memsw.failcnt
-rw-r--r--    1 root     root             0 Mar 12 02:28 memory.memsw.limit_in_bytes
-rw-r--r--    1 root     root             0 Mar 12 02:28 memory.memsw.max_usage_in_bytes
-r--r--r--    1 root     root             0 Mar 12 02:28 memory.memsw.usage_in_bytes
-rw-r--r--    1 root     root             0 Mar 12 02:28 memory.move_charge_at_immigrate
-r--r--r--    1 root     root             0 Mar 12 02:28 memory.numa_stat
-rw-r--r--    1 root     root             0 Mar 12 02:28 memory.oom_control
----------    1 root     root             0 Mar 12 02:28 memory.pressure_level
-rw-r--r--    1 root     root             0 Mar 12 02:28 memory.soft_limit_in_bytes
-r--r--r--    1 root     root             0 Mar 12 02:28 memory.stat
-rw-r--r--    1 root     root             0 Mar 12 02:28 memory.swappiness
-r--r--r--    1 root     root             0 Mar 12 02:28 memory.usage_in_bytes
-rw-r--r--    1 root     root             0 Mar 12 02:28 memory.use_hierarchy
-rw-r--r--    1 root     root             0 Mar 12 02:28 notify_on_release
-rw-r--r--    1 root     root             0 Mar 12 02:28 tasks
```

maybe other kubernetes users could validate; or someone with more experience with how cgroups work in kubernetes can comment?

**Comment by Tishj** [comment]
Fixed by #16608
